oeps/oep-0041-arch-async-server-event-messaging.rst

```diff
@@ -0,0 +1,486 @@
+* Real-time streaming of student learning events to external platforms, as
+  detailed in `OEP-26 <oep-0026-realtime-events>`_, though it is possible that
+  the implementation of OEP-26 might build on the infrastructure outlined in
+  this document.
+* Browser level notifications and event streaming along the lines of
+  Server-Sent-Events, GraphQL subscriptions, or frameworks built on top of
+  WebSocket. The interactions here are strictly server to server interactions
+  between services serving the same site. Again, it is entirely possible that
+  an implementation of browser level notifications would be built on top of the
+  infrastructure outlined here.
+
+
+----------
+Motivation
+----------
+
+Open edX has multiple services that need to copy data or take specific actions
+when certain events occur in other services. For example:
+
+* The start date of a course exists in both the Catalog domain represented by
+  the course-discovery service, as well as in the LMS, where its value is used
+  to help determine whether an enrolled student will have access to the content.
+* Other services often have a user table, with data replicated from the LMS
+  (the place where users register today).
+* Publishing a new version of course content in Studio triggers a variety of
+  LMS tasks to query and re-build optimized versions of course data for various
+  systems like the Block Transformers that power the mobile application and
+  course outline.
+* Grading events in the LMS may trigger certificate generation if the user has
+  exceeded the required threshold.
+
+Open edX services already use `Django signals
+<https://docs.djangoproject.com/en/1.11/topics/signals/>`_ for subscribing to
+events within the same service, but we have no accepted recommendation for how
+to do something similar across service boundaries. As the number of services in
+Open edX has grown, the various ad hoc synchronization and communication methods
+we've implemented over the years have caused numerous operational and
+performance problems (particularly around catalog data). At the same time,
+recent discussions around Domain Driven Design and data replication have given
+us a better framing for the kind of relationship we want to have between various
+services. There has also been a strong desire for a properly documented set of
+integration APIs that would allow Open edX to better interoperate with other
+applications in the ecosystem.
+
+There is an opportunity to bring all these elements together and create a solid
+foundation for simple, reliable, well documented, near real-time cross-service
+message passing. But without some planning and design work, we may end up in an
+even more chaotic place where each pairing of services uses different
+conventions for this kind of communication.
+
+
+-------------
+Specification
+-------------
+
+Message Format
+==============
+
+The message format will use the `CloudEvents 1.0
```

Comment:
my initial thoughts on the spec are: good. i think it’s worth adopting something other places are using to start with, and then we can make any additions/changes as we see fit 👍